### PR TITLE
T17 - String Permutations

### PR DESCRIPTION
**Pull request by vikrampg1710-tmm**
T17 - STRING PERMUTATIONS 

Generate all string character permutations. 

Input: "NOT", Output: "NOT NTO ONT OTN TNO TON"

### CODE REVIEW

File: Assignments/Program.cs [hunk 1]
```diff
@@ -1,3 +1,74 @@
...
+      input = input.ToUpper ();
+      if (input.Length == 2) return Permutated2 (input);
+      return Permutated34 (input);
+
+      /// <summary>Permutates given 2-letter word</summary>
+      static List<string> Permutated2 (string input)
+         => new () { $"{input[0]}{input[1]}", $"{input[1]}{input[0]}" };
+
+      /// <summary>Permutates given 3 [or] 4-letter word</summary>
+      static List<string> Permutated34 (string input) {
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs]: Instead of Permuted2 Permuted34 etc. can you try doing this with recursive method to do this?

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
Also do Edit > Advanced > Format Document. Unnecessary spaces found in the code.

File: Assignments/Program.cs [hunk 2]
```diff
@@ -1,3 +1,63 @@
...
+// Copyright (c) Metamation India.
+// ---------------------------------------------------------------------------------------
+// Program.cs
+// T16 - String Permutation
+// ---------------------------------------------------------------------------------------
 
-Console.WriteLine("Hello, World!");
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs:11]: image

File: Assignments/Program.cs [hunk 3]
```diff
@@ -1,3 +1,63 @@
...
+
+namespace Spark;
+public class T17 {
+   public static void Main () {
+      Console.WriteLine ("\x1B[4m" + "String Permutator:-" + "\x1B[0m");
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs]: image
Typos

File: Assignments/Program.cs [hunk 4]
```diff
@@ -1,3 +1,63 @@
...
+      }
+      foreach (char c in input) {
+         foreach (string s in store) {
+            if (!s.Contains (c)) {
+               string a = c + s;
+               if (!output.Contains (a)) output.Add (a);
+            }
+         }
+      }
+      return output;
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs]: Please re-check the logic. For the input word TAT, the output is not correct.
image

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Assignments/Program.cs (hunk 4 above)
Comment (vikrampg1710-tmm) [reply Assignments/Program.cs]: It cannot handle any strings with repeated chars.

Comment (varalakshmi-imm) [reply Assignments/Program.cs]: I expect tat, att, tta

File: Assignments/Program.cs [hunk 5]
```diff
@@ -1,3 +1,83 @@
...
+      }
+   }
+
+   /// <summary>Returns the list of permutated strings of given string</summary>
+   public static HashSet<string> PermutationsOf (string input) {
+      int len = input.Length;
+      HashSet<string> permutations = new ();
+      List<char> chars = new ();
+      for (int i = 0; i < len; i++)
+         chars.Add ((char)(48 + i));
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs]: Why should you convert to number and then to char?
Isn't this the index by the way?

File: Assignments/Program.cs [hunk 6]
```diff
@@ -1,3 +1,83 @@
...
+      var output = Permutated (new string (chars.ToArray ()));
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs]: output comes as numbers 012, 210 etc
Really finding it difficult to debug

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
Seems you are complicating the implementation.
Converting to index and back to char in the original input.
Not at all easy to debug.

Please read about KISS principle in software design.
Please write a simple implementation for this.

**Comment by varalakshmi-imm** [comment]
@vikrampg1710-tmm Did you read the review comments here https://github.com/vikrampg1710-tmm/Training/pull/17#pullrequestreview-1689787109?
Still the implementation is same char to indices and back to char after permutation except more lines of comments.
Though I respect the originality of your code, it looks too complicated.
Why should the input to your _Permutated_ method be 012 and not the original string NOT?

**Comment by vikrampg1710-tmm** [comment]
> @vikrampg1710-tmm Did you read the review comments here [#17 (review)](https://github.com/vikrampg1710-tmm/Training/pull/17#pullrequestreview-1689787109)? Still the implementation is same char to indices and back to char after permutation except more lines of comments. Though I respect the originality of your code, it looks too complicated. Why should the input to your _Permutated_ method be 012 and not the original string NOT?

The 'permutated' method cannot handle the input string with repeated chars in it.  So, I thought to permutate only the index values as a string (i.e., "012"), then using that to produce final permutated word.

But now, I modified the code in a such a way that it directly input the original string to the method, as you mentioned!

File: Assignments/Program.cs [hunk 7]
```diff
@@ -1,3 +1,65 @@
...
+      int count = 0;
+      foreach (var item in answer) {
+         Console.Write ($"{++count,3}. ");
+         Console.ForegroundColor = ConsoleColor.Green;
+         Console.WriteLine (item);
+         Console.ResetColor ();
+      }
+   }
+
+   /// <summary>A recusive method which returns permutations of given input word</summary>
```
Comment (varalakshmi-imm) [inline Assignments/Program.cs:39]:
```suggestion
   /// <summary>A recursive method which returns permutations of given input word</summary>

```
@vikrampg1710-tmm Check typos

**Review by varalakshmi-imm — APPROVED** [review]

**Pull request closed without merging**